File: src/libsokoengine/io/hexoban.cpp
#include "hexoban.hpp"

#include "rle.hpp"

#include <boost/algorithm/string.hpp>
#include <iostream>

using namespace std;

namespace sokoengine {
namespace io {

using game::Config;
using game::Tessellation;
using game::index_1d;
using game::X;
using game::Y;

namespace implementation {

class LIBSOKOENGINE_LOCAL HexobanPuzzleResizer : public PuzzleResizer {
public:
  virtual ~HexobanPuzzleResizer() = default;

  virtual void add_row_top(parsed_board_t &parsed_board, board_size_t &width,
                           board_size_t &height) const override;
  virtual void remove_row_top(parsed_board_t &parsed_board, board_size_t &width,
                              board_size_t &height) const override;
  virtual void remove_row_bottom(parsed_board_t &parsed_board, board_size_t &width,
                                 board_size_t &height) const override;
  virtual void reverse_columns(parsed_board_t &parsed_board, board_size_t &width,
                               board_size_t &height) const override;
};

class LIBSOKOENGINE_LOCAL HexobanPuzzleParser : public PuzzleParser {
public:
  virtual ~HexobanPuzzleParser() = default;
  virtual Strings parse(const string &board) const override;
};

class LIBSOKOENGINE_LOCAL HexobanPuzzlePrinter : public PuzzlePrinter {
public:
  virtual ~HexobanPuzzlePrinter() = default;
  virtual string print(const parsed_board_t &parsed_board, board_size_t width,
                       board_size_t height, bool use_visible_floor = false,
                       bool rle_encode = false) const override;
};

LIBSOKOENGINE_LOCAL const HexobanPuzzleResizer &hb_static_resizer() {
  static const HexobanPuzzleResizer the_one;
  return the_one;
}

LIBSOKOENGINE_LOCAL const HexobanPuzzleParser &hb_static_parser() {
  static const HexobanPuzzleParser the_one;
  return the_one;
}

LIBSOKOENGINE_LOCAL const HexobanPuzzlePrinter &hb_static_printer() {
  static const implementation::HexobanPuzzlePrinter the_one;
  return the_one;
}

} // namespace implementation

using namespace implementation;

class LIBSOKOENGINE_LOCAL HexobanPuzzle::PIMPL {
public:
  Snapshots m_snapshots;
};

HexobanPuzzle::HexobanPuzzle() : HexobanPuzzle(0, 0) {}

HexobanPuzzle::HexobanPuzzle(board_size_t width, board_size_t height)
  : Puzzle(Tessellation::HEXOBAN, hb_static_resizer(), hb_static_parser(),
           hb_static_printer(), width, height),
    m_impl(make_unique<PIMPL>()) {}

HexobanPuzzle::HexobanPuzzle(const string &src)
  : Puzzle(Tessellation::HEXOBAN, hb_static_resizer(), hb_static_parser(),
           hb_static_printer(), src),
    m_impl(make_unique<PIMPL>()) {}

HexobanPuzzle::HexobanPuzzle(const HexobanPuzzle &rv)
  : Puzzle(rv), m_impl(make_unique<PIMPL>(*rv.m_impl)) {}

HexobanPuzzle &HexobanPuzzle::operator=(const HexobanPuzzle &rv) {
  if (this != &rv) {
    Puzzle::operator=(rv);
    m_impl = make_unique<PIMPL>(*rv.m_impl);
  }
  return *this;
}

HexobanPuzzle::HexobanPuzzle(HexobanPuzzle &&rv) = default;

HexobanPuzzle &HexobanPuzzle::operator=(HexobanPuzzle &&rv) = default;

HexobanPuzzle::~HexobanPuzzle() = default;

HexobanPuzzle::unique_ptr_t HexobanPuzzle::clone() const {
  return make_unique<HexobanPuzzle>(*this);
}

const HexobanPuzzle::Snapshots &HexobanPuzzle::snapshots() const {
  return m_impl->m_snapshots;
}
HexobanPuzzle::Snapshots &HexobanPuzzle::snapshots() { return m_impl->m_snapshots; }

HexobanSnapshot::HexobanSnapshot() : Snapshot(Tessellation::HEXOBAN, "") {}

HexobanSnapshot::HexobanSnapshot(const string &moves_data)
  : Snapshot(Tessellation::HEXOBAN, moves_data) {}

HexobanSnapshot::HexobanSnapshot(const HexobanSnapshot &rv) : Snapshot(rv) {}

HexobanSnapshot &HexobanSnapshot::operator=(const HexobanSnapshot &rv) {
  if (this != &rv) { Snapshot::operator=(rv); }
  return *this;
}

HexobanSnapshot::HexobanSnapshot(HexobanSnapshot &&) = default;

HexobanSnapshot &HexobanSnapshot::operator=(HexobanSnapshot &&) = default;

HexobanSnapshot::~HexobanSnapshot() = default;

HexobanSnapshot::unique_ptr_t HexobanSnapshot::clone() const {
  return make_unique<HexobanSnapshot>(*this);
}

namespace implementation {

class LIBSOKOENGINE_LOCAL HexobanTextConverter {
  typedef tuple<Strings, board_size_t, board_size_t, int8_t, int8_t> preparse_results_t;
  typedef tuple<char, board_size_t, board_size_t, int8_t, int8_t>
    text_cell_position_data_t;
  typedef tuple<bool, bool> text_cell_position_status_t;

public:
  string convert_to_string(const parsed_board_t &parsed_board, board_size_t width,
                           board_size_t height, bool use_visible_floor = false) {
    char floor = use_visible_floor ? Puzzle::VISIBLE_FLOOR : Puzzle::FLOOR;

    Strings retv_list;
    for (position_t row = 0; row < height; row++) {
      string line;
      if (row % 2 != 0) {
        line += floor; // beginning half hex for odd rows
      }
      for (position_t col = 0; col < width; col++) {
        line += floor;

        char board_char = parsed_board[index_1d(col, row, width)];
        if (Puzzle::is_empty_floor(board_char)) {
          line += floor;
        } else {
          line += board_char;
        }
      }
      retv_list.push_back(line);
    }
    retv_list = PuzzleParser::normalize_width(retv_list, floor);
    if (is_type1(retv_list)) { remove_column_right(retv_list); }

    return boost::join(retv_list, "\n");
  }

  bool is_type1(const Strings &list) const {
    position_t rmnf = find_rightmost_non_floor(list);
    if (rmnf <= Config::MAX_POS) {
      board_size_t y = Y(rmnf, PuzzleParser::calculate_width(list));
      return y % 2 == 0;
    }
    return false;
  }

  pair<Strings, bool> convert_to_internal(const string &board) const {
    Strings parsed;
    board_size_t height, width;
    int8_t even_row_x_parity, odd_row_x_parity;

    tie(parsed, width, height, even_row_x_parity, odd_row_x_parity) =
      preparse_board(board);

    Strings internal_list;

    // Handle empty board
    if (width == 0 || height == 0) {
      return make_pair(Strings(), true);
    } else if (even_row_x_parity < 0 || odd_row_x_parity < 0) {
      for (board_size_t i = 0; i < height; i++) {
        internal_list.push_back(string(width / 2, Puzzle::VISIBLE_FLOOR));
      }
      return make_pair(internal_list, true);
    }

    bool layout_ok = true;

    for (board_size_t y = 0; y < height && layout_ok; y++) {
      string internal_line;
      for (board_size_t x = 0; x < width && layout_ok; x++) {
        bool should_copy_cell;

        tie(layout_ok, should_copy_cell) = analyze_text_cell_position(
          make_tuple(parsed[y][x], x, y, odd_row_x_parity, even_row_x_parity));

        if (layout_ok && should_copy_cell) internal_line += parsed[y][x];
      }
      if (layout_ok) internal_list.push_back(internal_line);
    }

    if (layout_ok)
      internal_list =
        PuzzleParser::normalize_width(internal_list, Puzzle::VISIBLE_FLOOR);

    return make_pair(internal_list, layout_ok);
  }

  void add_column_left(Strings &list) const {
    for (string &line : list)
      line.insert(line.cbegin(), Puzzle::VISIBLE_FLOOR);
  }

  void add_column_right(Strings &list) const {
    for (string &line : list)
      line.append(1, Puzzle::VISIBLE_FLOOR);
  }

  void add_row_top(Strings &list) const {
    list.insert(list.begin(), string(list.front().size(), Puzzle::VISIBLE_FLOOR));
  }

  void remove_column_right(Strings &list) const {
    for (string &line : list)
      line.pop_back();
  }

  void reverse_columns(Strings &list) const {
    for (string &line : list)
      std::reverse(line.begin(), line.end());
  }

  void remove_row_top(Strings &list) const { list.erase(list.begin()); }

  void remove_row_bottom(Strings &list) const { list.pop_back(); }

private:
  text_cell_position_status_t
  analyze_text_cell_position(text_cell_position_data_t position) const {
    char cell;
    board_size_t x, y;
    int8_t odd_row_x_parity, even_row_x_parity;
    tie(cell, x, y, odd_row_x_parity, even_row_x_parity) = position;

    int8_t y_parity = y % 2;
    int8_t x_parity = x % 2;

    bool layout_ok = true;
    // Is current cell part of board or only part of textual layout?
    bool is_cell_for_layout;
    bool should_copy_cell;

    // Check if textual encoding (layout) is legal. Positions of all board elements in
    // textual layout depend on position of first non floor element. If that element is
    // (odd column, even row) than all other elements in even rows must be in odd
    // columns. Other cells in even rows must be empty cells and their purpose is to
    // define board textual layout (they are not board elements).
    if (y_parity == 0) { // even rows
      if (x_parity == even_row_x_parity) {
        is_cell_for_layout = false;
      } else {
        // Cell is part of layout, it must be empty
        layout_ok = Puzzle::is_empty_floor(cell);
        is_cell_for_layout = true;
      }
    } else {        // odd rows
      if (x == 0) { // row start half hexes are always layout cells
        layout_ok = Puzzle::is_empty_floor(cell);
        is_cell_for_layout = true;
      } else {
        if (x_parity == odd_row_x_parity) {
          is_cell_for_layout = false;
        } else {
          // Cell is part of layout, it must be empty
          layout_ok = Puzzle::is_empty_floor(cell);
          is_cell_for_layout = true;
        }
      }
    }

    should_copy_cell = layout_ok && !is_cell_for_layout;

    return make_tuple(layout_ok, should_copy_cell);
  }

  preparse_results_t preparse_board(const string &board) const {
    Strings parsed;
    board_size_t height = 0, width = 0;

    parsed = PuzzleParser::cleaned_board_lines(board);
    height = parsed.size();
    width = height > 0 ? parsed.front().size() : 0;
    int8_t even_row_x_parity = -1, odd_row_x_parity = -1;

    if (height == 0 || width == 0)
      return make_tuple(parsed, width, height, even_row_x_parity, odd_row_x_parity);

    // Compensate for scheme2
    bool has_non_floor_left_in_odd_row = false;
    for (board_size_t i = 0; i < height && !has_non_floor_left_in_odd_row; ++i) {
      has_non_floor_left_in_odd_row =
        has_non_floor_left_in_odd_row ||
        (i % 2 == 1 && !Puzzle::is_empty_floor(parsed[i][0]));
    }
    if (has_non_floor_left_in_odd_row) {
      add_column_left(parsed);
      width += 1;
    }

    // Calculate parities
    position_t first_cell = find_first_non_floor(parsed);
    if (first_cell <= Config::MAX_POS) {
      int8_t first_cell_x_parity = X(first_cell, width) % 2;
      int8_t first_cell_y_parity = Y(first_cell, width) % 2;

      if (first_cell_y_parity == 0) {
        even_row_x_parity = first_cell_x_parity;
      } else {
        even_row_x_parity = (first_cell_x_parity + 1) % 2;
      }
      odd_row_x_parity = (even_row_x_parity + 1) % 2;
    }

    return make_tuple(parsed, width, height, even_row_x_parity, odd_row_x_parity);
  }

  position_t find_first_non_floor(const Strings &list) const {
    Strings normalized = HexobanPuzzleParser::normalize_width(list);
    board_size_t height = normalized.size();
    board_size_t width = height > 0 ? normalized.front().size() : 0;
    if (height == 0 || width == 0) return numeric_limits<position_t>::max();

    position_t x = 0, y = 0;
    bool non_floor_found = false;
    for (position_t row = 0; row < height && !non_floor_found; row++) {
      for (position_t col = 0; col < width && !non_floor_found; col++) {
        if (!Puzzle::is_empty_floor(normalized[row][col]) && (col > x || row > y)) {
          x = col;
          y = row;
          non_floor_found = true;
        }
      }
    }
    if (non_floor_found) {
      return index_1d(x, y, width);
    } else {
      return index_1d(1, 0, width); // Empty board, assuming scheme1
    }
  }

  position_t find_rightmost_non_floor(const Strings &strings) const {
    auto rightmost_finder = [](const Strings &strings,
                               int8_t row_parity) -> position_t {
      position_t retv = numeric_limits<position_t>::max();
      bool cell_found = false;
      board_size_t height = strings.size();
      board_size_t width = strings.front().size();
      position_t x = 0, y = 0;
      for (position_t row = row_parity % 2; row < height; row += 2) {
        for (position_t col = 0; col < width; col++) {
          if (!Puzzle::is_empty_floor(strings[row][col])) {
            cell_found = true;
            if (col > x || (col >= x && row > y)) {
              x = col;
              y = row;
            }
          }
        }
      }
      if (cell_found) { retv = index_1d(x, y, width); }
      return retv;
    };

    Strings normalized = HexobanPuzzleParser::normalize_width(strings);
    board_size_t height = normalized.size();
    board_size_t width = height > 0 ? normalized.front().size() : 0;
    if (height == 0 || width == 0) { return numeric_limits<position_t>::max(); }

    position_t rightmost_in_even_rows = rightmost_finder(normalized, 0);
    position_t rightmost_in_odd_rows = rightmost_finder(normalized, 1);

    if (rightmost_in_even_rows == numeric_limits<position_t>::max() ||
        rightmost_in_odd_rows == numeric_limits<position_t>::max()) { // Empty board
      return index_1d(0, 0, width);
    }

    position_t odd_x = X(rightmost_in_odd_rows, width);
    position_t odd_y = Y(rightmost_in_odd_rows, width);
    position_t even_x = X(rightmost_in_even_rows, width);
    position_t even_y = Y(rightmost_in_even_rows, width);

    if (odd_x > even_x) {
      return rightmost_in_odd_rows;
    } else if (even_x > odd_x) {
      return rightmost_in_even_rows;
    } else {
      return (odd_y >= even_y ? rightmost_in_odd_rows : rightmost_in_even_rows);
    }
  }

  bool is_type2(const Strings &strings) const { return !is_type1(strings); }
};

void HexobanPuzzleResizer::reverse_columns(parsed_board_t &parsed_board,
                                           board_size_t &width,
                                           board_size_t &height) const {
  HexobanTextConverter converter;

  Strings printed_board;
  boost::split(printed_board,
               converter.convert_to_string(parsed_board, width, height, true),
               boost::is_any_of("\n"));

  if (converter.is_type1(printed_board))
    converter.add_column_left(printed_board);
  else
    converter.add_column_right(printed_board);

  converter.reverse_columns(printed_board);
  converter.remove_column_right(printed_board);

  Strings new_parsed_board =
    converter.convert_to_internal(boost::join(printed_board, "\n")).first;
  _copy(parsed_board, width, height, new_parsed_board);
}

void HexobanPuzzleResizer::add_row_top(parsed_board_t &parsed_board,
                                       board_size_t &width,
                                       board_size_t &height) const {
  HexobanTextConverter converter;

  Strings printed_board;
  boost::split(printed_board,
               converter.convert_to_string(parsed_board, width, height, true),
               boost::is_any_of("\n"));

  converter.add_row_top(printed_board);

  Strings new_parsed_board =
    converter.convert_to_internal(boost::join(printed_board, "\n")).first;
  _copy(parsed_board, width, height, new_parsed_board);
}

void HexobanPuzzleResizer::remove_row_top(parsed_board_t &parsed_board,
                                          board_size_t &width,
                                          board_size_t &height) const {
  HexobanTextConverter converter;

  Strings printed_board;
  boost::split(printed_board,
               converter.convert_to_string(parsed_board, width, height, true),
               boost::is_any_of("\n"));

  converter.remove_row_top(printed_board);

  Strings new_parsed_board =
    converter.convert_to_internal(boost::join(printed_board, "\n")).first;
  _copy(parsed_board, width, height, new_parsed_board);
}

void HexobanPuzzleResizer::remove_row_bottom(parsed_board_t &parsed_board,
                                             board_size_t &width,
                                             board_size_t &height) const {
  HexobanTextConverter converter;

  Strings printed_board;
  boost::split(printed_board,
               converter.convert_to_string(parsed_board, width, height, true),
               boost::is_any_of("\n"));

  converter.remove_row_bottom(printed_board);

  Strings new_parsed_board =
    converter.convert_to_internal(boost::join(printed_board, "\n")).first;
  _copy(parsed_board, width, height, new_parsed_board);
}

Strings HexobanPuzzleParser::parse(const string &board) const {
  auto result = HexobanTextConverter().convert_to_internal(board);

  if (result.second) return result.first;

  throw invalid_argument(
    "String can't be parsed to HexobanPuzzle. Probable cause is invalid "
    "text layout meaning either missing or misaligned filler floor "
    "characters.");
}

string HexobanPuzzlePrinter::print(const parsed_board_t &parsed_board,
                                   board_size_t width, board_size_t height,
                                   bool use_visible_floor, bool rle_encode) const {
  string retv = HexobanTextConverter().convert_to_string(parsed_board, width, height,
                                                         use_visible_floor);

  if (rle_encode) { retv = Rle::encode(retv); }

  return retv;
}

} // namespace implementation
} // namespace io
} // namespace sokoengine

File: src/sokoenginepy/io/hexoban.py
from __future__ import annotations

from typing import List, Optional, Tuple

from .puzzle import Puzzle
from .puzzle_parsing import PuzzleParser, PuzzlePrinter, PuzzleResizer
from .rle import Rle
from .snapshot import Snapshot


class HexobanSnapshot(Snapshot):
    def __init__(self, moves_data: str = ""):
        from ..game import Tessellation

        super().__init__(tessellation=Tessellation.HEXOBAN, moves_data=moves_data)


class HexobanPuzzle(Puzzle):
    def __init__(self, width: int = 0, height: int = 0, board: Optional[str] = None):
        """
        Arguments:
            width: number of columns
            height: number of rows
            board: If not blank, it will be parsed and board will be created from it,
                ignoring ``width`` and ``height``.
        """
        from ..game import Tessellation

        super().__init__(
            tessellation=Tessellation.HEXOBAN,
            width=width,
            height=height,
            board=board,
            resizer_cls=HexobanPuzzleResizer,
            parser_cls=HexobanPuzzleParser,
            printer_cls=HexobanPuzzlePrinter,
        )


class HexobanTextConverter:
    def convert_to_string(
        self,
        parsed_board: List[str],
        width: int,
        height: int,
        use_visible_floor: bool,
    ) -> str:
        from ..game import index_1d

        floor = Puzzle.VISIBLE_FLOOR if use_visible_floor else Puzzle.FLOOR

        retv: List[str] = []
        for row in range(0, height):
            line = ""
            if row % 2 == 1:
                # beginning half hex for odd rows
                line += floor

            for col in range(0, width):
                puzzle_char = parsed_board[index_1d(col, row, width)]

                line += floor
                if Puzzle.is_empty_floor(puzzle_char):
                    line += floor
                else:
                    line += puzzle_char

            retv.append(line)

        retv = PuzzleParser.normalize_width(retv, floor)
        if self.is_type1(retv):
            retv = self.remove_column_right(retv)

        return "\n".join(retv)

    def is_type1(self, strings: List[str]) -> bool:
        from ..game import Y

        rmnf = self._find_rightmost_non_floor(strings)

        if rmnf is None:
            return False
        elif rmnf > 0:
            y = Y(rmnf, PuzzleParser.calculate_width(strings))
            return y % 2 == 0
        else:  # rmnf == 0
            return True

    @staticmethod
    def _find_rightmost_non_floor(strings: List[str]) -> Optional[int]:
        from ..game import X, Y, index_1d

        def rightmost_finder(strings: List[str], row_parity: int):
            cell_found = False
            height = len(strings)
            width = len(strings[0])
            x = y = 0

            for row in range(row_parity % 2, height, 2):
                for col in range(0, width):
                    if not Puzzle.is_empty_floor(strings[row][col]):
                        cell_found = True
                        if col > x or (col >= x and row > y):
                            x = col
                            y = row

            if cell_found:
                return index_1d(x, y, width)

            return None

        normalized = HexobanPuzzleParser.normalize_width(strings)
        width = len(normalized[0]) if normalized else 0
        height = len(normalized)

        if width == 0 or height == 0:
            return None

        rightmost_in_even_rows = rightmost_finder(normalized, 0)
        rightmost_in_odd_rows = rightmost_finder(normalized, 1)

        if rightmost_in_even_rows is None or rightmost_in_odd_rows is None:
            return index_1d(0, 0, width)

        odd_x = X(rightmost_in_odd_rows, width)
        odd_y = Y(rightmost_in_odd_rows, width)
        even_x = X(rightmost_in_even_rows, width)
        even_y = Y(rightmost_in_even_rows, width)

        if odd_x > even_x:
            return rightmost_in_odd_rows
        elif even_x > odd_x:
            return rightmost_in_even_rows

        return rightmost_in_odd_rows if odd_y > even_y else rightmost_in_even_rows

    def convert_to_internal(self, board: str) -> Tuple[List[str], bool]:
        # Converts textual Hexoban into 2D array and validates textual layout
        (
            parsed,
            width,
            height,
            even_row_x_parity,
            odd_row_x_parity,
        ) = self._preparse_board(board)

        # Handle empty board
        if width == 0 or height == 0:
            return [], True

        elif even_row_x_parity < 0 or odd_row_x_parity < 0:
            internal = height * [int(width / 2) * Puzzle.VISIBLE_FLOOR]
            return internal, True

        layout_ok = True
        internal: List[str] = []

        for y in range(0, height):
            if not layout_ok:
                break
            internal_line = ""
            for x in range(0, width):
                if not layout_ok:
                    break

                layout_ok, should_copy_cell = self._analyze_text_cell_position(
                    parsed[y][x], x, y, odd_row_x_parity, even_row_x_parity
                )

                if layout_ok and should_copy_cell:
                    internal_line += parsed[y][x]

            if layout_ok:
                internal.append(internal_line)

        if layout_ok:
            internal = PuzzleParser.normalize_width(internal, Puzzle.VISIBLE_FLOOR)

        return internal, layout_ok

    def _is_type2(self, strings: List[str]) -> bool:
        return not self.is_type1(strings)

    @staticmethod
    def _analyze_text_cell_position(
        cell: str, x: int, y: int, odd_row_x_parity: int, even_row_x_parity: int
    ) -> Tuple[bool, bool]:
        y_parity = y % 2
        x_parity = x % 2

        layout_ok = True
        # Is current cell part of board or only part of textual layout?
        is_cell_for_layout = False
        should_copy_cell = False

        # Check if textual encoding (layout) is legal. Positions of all
        # board elements in textual layout depend on position of first
        # non floor element. If that element is (odd column, even row)
        # than all other element in even rows must be in odd columns.
        # Other cells in even rows must be empty cells and their purpose
        # is to define board textual layout (they are not board elements).
        if y_parity == 0:  # even rows
            if x_parity == even_row_x_parity:
                is_cell_for_layout = False
            else:  # Cell is part of layout, it must be empty
                layout_ok = Puzzle.is_empty_floor(cell)
                is_cell_for_layout = True
        else:  # odd rows
            if x == 0:  # row start half hexes are always layout cells
                layout_ok = Puzzle.is_empty_floor(cell)
                is_cell_for_layout = True
            else:
                if x_parity == odd_row_x_parity:
                    is_cell_for_layout = False
                else:  # Cell is part of layout, it must be empty
                    layout_ok = Puzzle.is_empty_floor(cell)
                    is_cell_for_layout = True

        should_copy_cell = layout_ok and not is_cell_for_layout

        return layout_ok, should_copy_cell

    def _preparse_board(self, board: str) -> Tuple[List[str], int, int, int, int]:
        from ..game import X, Y

        parsed: List[str] = PuzzleParser.cleaned_board_lines(board)
        width = len(parsed[0]) if parsed else 0
        height = len(parsed)

        even_row_x_parity = odd_row_x_parity = -1

        if height != 0 and width != 0:
            # Compensate for scheme2
            has_non_floor_left_in_odd_row = False

            for i in range(0, height):
                if has_non_floor_left_in_odd_row:
                    break
                has_non_floor_left_in_odd_row = has_non_floor_left_in_odd_row or (
                    i % 2 == 1 and not Puzzle.is_empty_floor(parsed[i][0])
                )

            if has_non_floor_left_in_odd_row:
                parsed = self.add_column_left(parsed)
                width += 1

            # Calculate parities
            first_cell = self._find_first_non_floor(parsed)
            if first_cell is not None:
                first_cell_x_parity = X(first_cell, width) % 2
                first_cell_y_parity = Y(first_cell, width) % 2

                if first_cell_y_parity == 0:
                    even_row_x_parity = first_cell_x_parity
                else:
                    even_row_x_parity = (first_cell_x_parity + 1) % 2

                odd_row_x_parity = (even_row_x_parity + 1) % 2

        return parsed, width, height, even_row_x_parity, odd_row_x_parity

    @staticmethod
    def add_column_left(strings: List[str]) -> List[str]:
        return [Puzzle.VISIBLE_FLOOR + line for line in strings]

    @staticmethod
    def add_column_right(strings: List[str]) -> List[str]:
        return [line + Puzzle.VISIBLE_FLOOR for line in strings]

    @staticmethod
    def add_row_top(strings: List[str]) -> List[str]:
        w = HexobanPuzzleParser.calculate_width(strings)
        return [w * Puzzle.VISIBLE_FLOOR] + strings

    @staticmethod
    def remove_column_right(strings: List[str]) -> List[str]:
        return [line[:-1] for line in strings]

    @staticmethod
    def reverse_columns(strings: List[str]) -> List[str]:
        return ["".join(c for c in reversed(line)) for line in strings]

    @staticmethod
    def remove_row_top(strings: List[str]) -> List[str]:
        return strings[1:]

    @staticmethod
    def remove_row_bottom(strings: List[str]) -> List[str]:
        return strings[:-1]

    @staticmethod
    def _find_first_non_floor(strings: List[str]):
        from ..game import index_1d

        normalized = HexobanPuzzleParser.normalize_width(strings)
        width = len(normalized[0]) if normalized else 0
        height = len(normalized)

        if width == 0 or height == 0:
            return None

        x = y = 0
        non_floor_found = False

        for row in range(0, height):
            if non_floor_found:
                break
            for column in range(0, width):
                if non_floor_found:
                    break

                if not Puzzle.is_empty_floor(normalized[row][column]) and (
                    column > x or row > y
                ):
                    x = column
                    y = row
                    non_floor_found = True

        if non_floor_found:
            return index_1d(x, y, width)

        # Empty board, assuming scheme1
        return index_1d(1, 0, width)


class HexobanPuzzlePrinter(PuzzlePrinter):
    def print(
        self,
        parsed_board: List[str],
        width: int,
        height: int,
        use_visible_floor: bool,
        rle_encode: bool,
    ) -> str:
        retv = HexobanTextConverter().convert_to_string(
            parsed_board, width, height, use_visible_floor
        )
        if rle_encode:
            return Rle.encode(retv)
        return retv


class HexobanPuzzleParser(PuzzleParser):
    def parse(self, board: str) -> List[str]:
        parsed, layout_ok = HexobanTextConverter().convert_to_internal(board)

        if layout_ok:
            return parsed

        raise ValueError(
            "String can't be parsed to HexobanPuzzle. Probable cause is invalid "
            "text layout meaning either missing or misaligned filler floor "
            "characters."
        )


class HexobanPuzzleResizer(PuzzleResizer):
    def reverse_columns(
        self, parsed_board: List[str], width: int, height: int
    ) -> Tuple[List[str], int, int]:
        converter = HexobanTextConverter()

        printed_board = converter.convert_to_string(
            parsed_board, width, height, True
        ).splitlines()

        if converter.is_type1(printed_board):
            printed_board = converter.add_column_left(printed_board)
        else:
            printed_board = converter.add_column_right(printed_board)

        printed_board = converter.reverse_columns(printed_board)
        printed_board = converter.remove_column_right(printed_board)

        new_parsed_board, layout_ok = converter.convert_to_internal(
            "\n".join(printed_board)
        )
        new_width = len(new_parsed_board[0]) if new_parsed_board else 0
        new_height = len(new_parsed_board)

        return list("".join(new_parsed_board)), new_width, new_height

    def add_row_top(
        self, parsed_board: List[str], width: int, height: int
    ) -> Tuple[List[str], int, int]:
        converter = HexobanTextConverter()

        printed_board = converter.convert_to_string(
            parsed_board, width, height, True
        ).splitlines()

        printed_board = converter.add_row_top(printed_board)

        new_parsed_board, layout_ok = converter.convert_to_internal(
            "\n".join(printed_board)
        )
        new_width = len(new_parsed_board[0]) if new_parsed_board else 0
        new_height = len(new_parsed_board)

        return list("".join(new_parsed_board)), new_width, new_height

    def remove_row_top(
        self, parsed_board: List[str], width: int, height: int
    ) -> Tuple[List[str], int, int]:
        converter = HexobanTextConverter()

        printed_board = converter.convert_to_string(
            parsed_board, width, height, True
        ).splitlines()

        printed_board = converter.remove_row_top(printed_board)

        new_parsed_board, layout_ok = converter.convert_to_internal(
            "\n".join(printed_board)
        )
        new_width = len(new_parsed_board[0]) if new_parsed_board else 0
        new_height = len(new_parsed_board)

        return list("".join(new_parsed_board)), new_width, new_height

    def remove_row_bottom(
        self, parsed_board: List[str], width: int, height: int
    ) -> Tuple[List[str], int, int]:
        converter = HexobanTextConverter()

        printed_board = converter.convert_to_string(
            parsed_board, width, height, True
        ).splitlines()

        printed_board = converter.remove_row_bottom(printed_board)

        new_parsed_board, layout_ok = converter.convert_to_internal(
            "\n".join(printed_board)
        )
        new_width = len(new_parsed_board[0]) if new_parsed_board else 0
        new_height = len(new_parsed_board)

        return list("".join(new_parsed_board)), new_width, new_height
